[WIP] UI: Move bookmark button to primary navigation area

petemill opened the pull request:
## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.

## Test Plan:


## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions

petemill commented:
Now rebased on latest master with C69.

petemill commented:
Another question for reviewers: could  / should `BraveBrowserView::SetStarredState` be re-defined in using the `chromium_src` directory technique instead of a subclass / override?

bbondy commented on patches/chrome-browser-ui-views-frame-browser_window_factory.cc.patch:
```diff
@@ -0,0 +1,22 @@
+   // Create the view and the frame. The frame will attach itself via the view
+   // so we don't need to do anything with the pointer.
+-  BrowserView* view = new BrowserView();
++  BrowserView* view = new BraveBrowserView();
```
This can be done via chromium_src override.
So create a file in src/brave/chrome/browser/ui/views/frame/browser_window_factory.cc

At the top of the file include `#include "chrome/browser/ui/views/frame/browser_view.h"` s that it doesn't get renamed too.

Then

#define BrowserView BraveBrowserView
Then `#include "../../../../../../../chrome/browser/ui/views/frame/browser_window_factory.cc"`

bbondy commented on patches/chrome-browser-ui-views-location_bar-location_bar_view.cc.patch:
```diff
@@ -0,0 +1,17 @@
++#if !defined(BRAVE_CHROMIUM_BUILD)
+   if (browser_) {
+     page_action_icons_.push_back(
+         star_view_ = new StarView(command_updater(), browser_, this));
```
I'm not sure if it's possible but I wonder if we did chromium_src override and replaced StarView with something else like NullView. Check if you can modify StarView to do nothing, and if so make a new NullView that you replace it with without patching.

bbondy commented on patches/chrome-browser-ui-views-toolbar-toolbar_view.cc.patch:
```diff
@@ -0,0 +1,72 @@
+ #include "ui/views/widget/tooltip_manager.h"
+ #include "ui/views/widget/widget.h"
+ #include "ui/views/window/non_client_view.h"
++#include "brave/browser/ui/views/toolbar/bookmark_button.h"
```
This patching is very bad in this file but I don't have a good idea on how to improve it.

bridiver replied:
I think most likely by subclassing

bridiver replied:
but actually in this case I think we want to just replace LocationBarView. We'll remove the original StarView using the NullView you suggested and then create a new view that adds the LocationBarView and the StarView/Bookmark Button as child views

petemill replied:
@bridiver that's an interesting idea. Some concerns before I proceed with it that I'd appreciate your thoughts on:

- `location_bar` is a public accessibly member of `ToolbarView`. Is there an easy way to make our new `BookmarkButtonAndLocationBarView` pass through all direct calls to public `LocationBarView` members to its own actual `location_bar` member?

- When layout out the view in the x axis, chromium uses the value `TOOLBAR_ELEMENT_PADDING` for the gap between each icon, but the value `TOOLBAR_STANDARD_SPACING` for the gap between the last icon (home, reload, or bookmark in Brave's case) and the location bar. If we inserted the icon in to the left of the brave location bar, then it would have incorrect spacing. If `TOOLBAR_STANDARD_SPACING` is less than `TOOLBAR_ELEMENT_PADDING` then we could add in the difference, but it feel like a hack.

- This is a (visually) trivial first of many non-trivial layout / design changes, so we may have to subclass `BrowserView` anyway. Though of course better to approach case-by-case.

bbondy commented on patches/chrome-browser-ui-views-bookmarks-bookmark_bubble_view.cc.patch:
```diff
@@ -0,0 +1,13 @@
+   // LocationBarBubbleDelegateView does not know that and may use different
+   // arrow anchoring.
+-  bookmark_bubble_->set_arrow(views::BubbleBorder::TOP_RIGHT);
++  bookmark_bubble_->set_arrow(views::BubbleBorder::TOP_LEFT);
```
I think you can use chromium_src override here and rename `TOP_RIGHT` to `TOP_LEFT` w/ `#define`
You'll need some includes at the top as usual from `chrome/browser/ui/views/bookmarks/bookmark_bubble_view.cc` inside the `chromium_src` override.

bridiver replied:
I'm more concerned about the comment above it

bridiver replied:
// Bookmark bubble should always anchor TOP_RIGHT, but the
  // LocationBarBubbleDelegateView does not know that and may use different
  // arrow anchoring

bridiver replied:
so it seems like we shouldn't change this and also that there is another way to set it to LEFT

petemill replied:
@bridiver I think that comment is saying that `LocationBarBubbleDelegateView` may have another default, or may calculate anchor point based on window width, etc, but for the Bookmark version of that delegate it should always be `TOP_RIGHT` for Chrome (or `TOP_LEFT` for Brave), which would mean this would be the correct place to specify this. We do not want to modify the anchor point for any other bubble views.

@bbondy I'll try the chromium_src method

petemill replied:
As discussed with @bridiver, we cannot redefine the enum value using the pre-processor. So instead I tried to re-call `set_arrow` in our own code via the following:
```
bubble_widget->widget_delegate()->AsBubbleDialogDelegate()->set_arrow(views::BubbleBorder::TOP_LEFT);
```
But this value is used during `Init`, which is called during `CreateBubble`, which has already been called by the time the bubble is returned to us.

So the only other option would be to override `BookmarkBubbleView::ShowBubble` but there is quite a bit of functional logic in that method, which seems more risky than this small patch.

bbondy commented on browser/ui/views/toolbar/bookmark_button.cc:
```diff
@@ -0,0 +1,98 @@
+                    ui::NativeTheme::kColorId_ProminentButtonColor)
+          : tp->GetColor(ThemeProperties::COLOR_TOOLBAR_BUTTON_ICON);
+  SetImage(views::Button::STATE_NORMAL, gfx::CreateVectorIcon(
+      toolbar::kStarIcon, icon_color));
```
Maybe we can just use our own vector icon here instead of the patch below to use his one?

bridiver replied:
I'm a little confused about this whole class because the description sounds like we are just moving it, so why do we need a new class? Can't we just re-use the existing StarView in a different place?

petemill replied:
@bbondy we are not patching the icon - it's our own icon file, I just needed to add a different size variant to the file.

@bridiver `StarView` inherits `PageActionIconView`, which expects to be inside the `LocationBarView` visually and hierarchically. My first attempt at this did re-use that class, but it came with some side effects (mainly visual). The visual aim is to have this be another top-level toolbar button, which `ToolbarButton` achieves without any extra rendering code. The modifications to `PageActionIconView` may have been numerable and would be repeating visual style, making it potentially more difficult to rebase in the future than using the simple `ToolbarButton` class.

simonhong replied:
How about adding new `UpdateIconImage()` with codes from line 60 to last and calling it when theme changing event occurs because this button's color depends on theme property.

bridiver commented on patches/chrome-browser-ui-views-toolbar-toolbar_view.cc.patch:
```diff
@@ -0,0 +1,72 @@
+   views::View* anchor_view = location_bar();
++#if !defined(BRAVE_CHROMIUM_BUILD)
+   PageActionIconView* const star_view = location_bar()->star_view();
+   if (!ui::MaterialDesignController::IsSecondaryUiMaterial()) {
```
this seems to be controlled by IsSecondaryUiMaterial which I believe is controllable by a flag

petemill replied:
@bridiver yes and that flag is true by default at the moment as far as I know, so this code block shouldn't get run anyway. However, we still need the `PageActionIconView* const star_view = location_bar()->star_view();` line not to run. However, if we follow your suggestion to replace `location_bar` with a class that encompasses the `BookmarkButton` and the actual `LocationBarView` then this may not be necessary.

petemill commented:
Thanks for the review @bbondy and @bridiver.

In this PR, I’ve addressed some of the feedback in order to reduce patching, notably subclassing ToolbarView and LocationBarView.
However:
- For now, this copies some of the method implementations we’re only slightly modifying, such as `Toolbar::Layout`. However, I think this could be a valid way since the (more-)finalized designs now incorporate more layout changes, potentially meaning more rewrite of `ToolbarView::Layout` (and `ToolbarView:GetSizeInternal`).
- I still had to patch to change a few members to `virtual` and make a `friend` declaration.
- It’s not so immediately obvious to me how to use chromium_src to rewrite a enum value, in this case `BubbleBorder::Arrow::TOP_RIGHT` at https://cs.chromium.org/chromium/src/ui/views/bubble/bubble_border.h?rcl=6048aabe0b5a5f1515bcc88a2aff19ce8ccaa5b5&l=85

I've kept that ^ as a separate commit, in case we want to instead follow one of @bridiver's suggestion of a `LocationBarWrapperView` class that `ToolbarView` will use, and have `LocationBarView` and `BookmarkButton` as children. I spent a quite bit of time working on that option but it does not seem as preferrable as the subclass option in this PR to me. Before getting too far, I wanted to get some feedback on if this is really a good way forward because:

- There are many references up the tree to toolbar_view()->location_bar() (expecting that to be both a LocationBarView and LocationBar. This means both implementing many many members, and also rewriting type references in many files, including cocoa.
    - This can be patched in the getter to return BraveLocationBarWrapperView->LocationBarView instead, which would preserve the types, but we would actually want at least one of those calls to refer to the wrapper (`SetStarToggled`) and the rest to refer to the child, so that doesn't work.
    - We could instead make the wrapper *inherit* from LocationBarView and LocationBar, instead of copy the members, but that could have bad side-effects / performance.

- It still likely requires subclassing `ToolbarView` to get rid of all patching because the added `BookmarkButton` is not a `PageActionIconView`, it's a `ToolbarButton` - `ToolbarView::ShowBookmarkBubble` declares a variable referencing the star_icon with that type.
- `ToolbarView::Layout` sets a specific height on `location_bar_` which may not be tall enough for the button.

I've pushed a branch with this option so far. It's not compiling yet due to the high number of references to LocationBarView that need to be rewritten (for now I started to do them as patches rather than chromium_src): https://github.com/petemill/brave-core/commit/bc2f4d92767f29ee374f4b2d6f88ec56e9fee657

petemill commented:
This PR is ready for re-review @bbondy @bridiver 

After a DM conversation with @bridiver it's been decided that subclassing and copying / modifying Layout code is fine as we'll most likely be changing it, and should be treated differently than copying functional code to a subclass.

So the patching is now minimized simply to making some members `virtual`, as well as the `BookmarkBubbleView`'s anchor alignment (left -> right) which seems unavoidable (comment left inline with the approaches I tried ^).

simonhong commented on browser/ui/views/toolbar/brave_toolbar_view.cc:
```diff
@@ -0,0 +1,206 @@
+    return;
+  }
+
+  bookmark_ = new BookmarkButton(this);
```
Don't we need to apply upstream's button visibility condition like below?
```
oid LocationBarView::UpdateBookmarkStarVisibility() {
  if (star_view_) {
    star_view_->SetVisible(
        browser_defaults::bookmarks_enabled && !is_popup_mode_ &&
        !GetToolbarModel()->input_in_progress() &&
        edit_bookmarks_enabled_.GetValue() &&
        !IsBookmarkStarHiddenByExtension());
  }
}
```

petemill replied:
@simonhong I had considered it but I didn't think we needed to because:

- `popup_mode`: this is taken care of in the ToolbarView which will only render the `LocationBarView` when `popup_mode = true`. Since the `BookmarkButton` isn't in the `LocationBarView` anymore, it doesn't need to consider it anymore.
- `input_in_progress`: again, this only makes sense when the bookmark button is inside the location bar, which it isn't anymore
- `IsBookmarkStarHiddenByExtension`: I believe that is for the following extension, which is being discontinued. I don't believe it's an officially supported extension option, but I could be wrong?https://chrome.google.com/webstore/detail/bookmark-manager/gmlllbghnfkpflemihljekbapjopfjik
- `bookmarks_enabled` and `edit_bookmarks_enabled_`: brave-core will always have bookmarks enabled. However, chromium's "Guest Mode" windows do not allow bookmarking. According to the Product team, I believe we won't have Guest Mode windows. It may be worth supporting this just in case, but I was trying to avoid having conditional layout for the button existence.

Let me know if you have any thoughts on those items @simonhong

simonhong replied:
@petemill , got it. Thanks for kind explanation!

petemill replied:
Now caters to `bookmarks_enabled` and `edit_bookmarks_enabled_` for guest windows.

simonhong commented on browser/ui/views/toolbar/brave_toolbar_view.cc:
```diff
@@ -0,0 +1,206 @@
+  // (i.e. in non-normal display mode).
+  views::View* anchor_view = location_bar();
+  BookmarkButton* const star_view = bookmark_button();
+  if (star_view && star_view->visible())
```
nit: `star_view->visible()` can be removed because it is always visible.

petemill replied:
no longer true since `visible` is `false` when bookmarks are disabled.

petemill commented:
This is ready now IMO.

Recent changes:
- Support modes where bookmarking is not enabled (e.g. Guest Window)
- 'on' version of icon is filled-in grey, instead of outlined [primary].

simonhong commented:
Do you think that bookmark button doesn't need to care about theme change? 
https://github.com/brave/brave-core/pull/280#discussion_r209486496

petemill commented:
@simonhong I don't believe it's necessary. I just tested changing the theme, and this button updates with the new theme color correctly, same as the other buttons which inherit from `ToolbarButton`.

simonhong commented:
Ah, it's good then. I assume that `SetToggled(bool on)` is called from somewhere else when theme is changed. :)

simonhong commented:
@bbondy @bridiver Any more comments about this PR? I think this PR is ready to merge
